<script>
  Promise.resolve = function (x) {
    return new Promise((resolve) => {
      if (x.then) {
        x.then((res) => {
          console.log("rrrr", res);
          Promise.resolve(res);
        });
      } else {
        resolve(x);
      }
    });
  };
  Promise.resolve(
    new Promise((resolve, reject) =>
      setTimeout(console.log.bind(null, 123), 2000)
    )
  )
    .then(console.log)
    .catch(console.log);

  var mp = Promise.resolve(333)
    .then((res) => {
      console.log(res);
      return 666;
    })
    .then(console.log);

  Promise.resolve(999).then((res) => {
    console.log(res);
  });

  console.log("normal");
  setTimeout(() => {
    console.log("timeout");
  }, 0);
</script>
<script>
  class Promise {
    constructor(executor) {
      this.state = "pending";
      this.value = undefined;
      this.reason = undefined;
      this.onResolvedCallbacks = [];
      this.onRejectedCallbacks = [];
      let resolve = (value) => {
        if (this.state === "pending") {
          this.state = "fulfilled";
          this.value = value;
          this.onResolvedCallbacks.forEach((fn) => fn());
        }
      };
      let reject = (reason) => {
        if (this.state === "pending") {
          this.state = "rejected";
          this.reason = reason;
          this.onRejectedCallbacks.forEach((fn) => fn());
        }
      };
      try {
        executor(resolve, reject);
      } catch (err) {
        reject(err);
      }
    }
    then(onFulfilled, onRejected) {
      onFulfilled =
        typeof onFulfilled === "function" ? onFulfilled : (value) => value;
      onRejected =
        typeof onRejected === "function"
          ? onRejected
          : (err) => {
              throw err;
            };

      // 1. promise2是必须返回的
      // 把 then 的返回值用 promise 套上：
      let promise2 = new Promise((resolve, reject) => {
        if (this.state === "fulfilled") {
          setTimeout(() => {
            try {
              let x = onFulfilled(this.value);
              resolvePromise(promise2, x, resolve, reject);
            } catch (e) {
              reject(e);
            }
          }, 0);
        }
        if (this.state === "rejected") {
          setTimeout(() => {
            try {
              let x = onRejected(this.reason);
              resolvePromise(promise2, x, resolve, reject);
            } catch (e) {
              reject(e);
            }
          }, 0);
        }
        if (this.state === "pending") {
          this.onResolvedCallbacks.push(() => {
            setTimeout(() => {
              try {
                let x = onFulfilled(this.value);
                resolvePromise(promise2, x, resolve, reject);
              } catch (e) {
                reject(e);
              }
            }, 0);
          });
          this.onRejectedCallbacks.push(() => {
            setTimeout(() => {
              try {
                let x = onRejected(this.reason);
                resolvePromise(promise2, x, resolve, reject);
              } catch (e) {
                reject(e);
              }
            }, 0);
          });
        }
      });
      return promise2;
    }
    catch(fn) {
      return this.then(null, fn);
    }
  }
  function resolvePromise(promise2, x, resolve, reject) {
    if (x === promise2) {
      return reject(new TypeError("Chaining cycle detected for promise"));
    }
    let called;
    if (x != null && (typeof x === "object" || typeof x === "function")) {
      // 如果不为空并且是对象或函数：
      try {
        let then = x.then;
        // 尝试拿到then返回值x的then方法，如果有then方法：
        if (typeof then === "function") {
          //   接收到thenable类型的对象的时候：
          //      返回一个新的promise对象，这个对象具有一个 then 方法

          // 调用then方法？？？
          then.call(
            // 如果x是promise，则取他的结果（通过then的参数获取）作为promise2的成功结果
            x,
            (y) => {
              // 拿到then方法resolve的结果传给
              if (called) return;
              called = true;
              resolvePromise(promise2, y, resolve, reject);
            },
            (err) => {
              if (called) return;
              called = true;
              reject(err);
            }
          );
        } else {
          resolve(x);
        }
      } catch (e) {
        if (called) return;
        called = true;
        reject(e);
      }
    } else {
      resolve(x);
    }
  }
  //resolve方法
  Promise.resolve = function (val) {
    return new Promise((resolve, reject) => {
      resolve(val);
    });
  };
  //reject方法
  Promise.reject = function (val) {
    return new Promise((resolve, reject) => {
      reject(val);
    });
  };
  //race方法
  Promise.race = function (promises) {
    return new Promise((resolve, reject) => {
      for (let i = 0; i < promises.length; i++) {
        promises[i].then(resolve, reject);
      }
    });
  };
  //all方法(获取所有的promise，都执行then，把结果放到数组，一起返回)
  Promise.all = function (promises) {
    let arr = [];
    let i = 0;
    function processData(index, data) {
      arr[index] = data;
      i++;
      if (i == promises.length) {
        resolve(arr);
      }
    }
    return new Promise((resolve, reject) => {
      for (let i = 0; i < promises.length; i++) {
        promises[i].then((data) => {
          processData(i, data);
        }, reject);
      }
    });
  };

  const p = new Promise((resolve) =>
    setTimeout(() => {
      resolve("timeout");
    }, 2000)
  ).then((res) => {
    console.log("res", res);
    return new Promise((resolve) =>
      setTimeout(() => {
        resolve("timeout2");
      }, 1000)
    );
  });
</script>
